tianshou/data/utils/converter.py

```diff
@@ -80,3 +82,62 @@ def to_torch_as(
     """
     assert isinstance(y, torch.Tensor)
     return to_torch(x, dtype=y.dtype, device=y.device)
+
+# Note: object is used as a proxy for objects that can be pickled
+Hdf5ConvertibleValues = Union[
+    int, float, Batch, np.ndarray, torch.Tensor, object, 'Hdf5ConvertibleType'
+]
+Hdf5ConvertibleType = Dict[str, Hdf5ConvertibleValues]
+
+def to_hdf5(x: Hdf5ConvertibleType, y: h5py.Group) -> None:
+    """Copy object into HDF5 group."""
+    for k, v in x.items():
+        # dicts and batches are both represented by groups
+        if isinstance(v, (Batch, dict)):
+            subgrp = y.create_group(k)
+            if isinstance(v, dict):
+                subgrp_data = v
+            else:
+                subgrp_data = v.__getstate__()
+                subgrp.attrs["__convert_to__"] = "Batch"
+            to_hdf5(subgrp_data, subgrp)
+        # numpy arrays and pytorch tensors are written to datasets
+        elif isinstance(v, np.ndarray):
+            y.create_dataset(k, data=v)
+            y[k].attrs["__data_type__"] = "numpy"
+        elif isinstance(v, torch.Tensor):
+            y.create_dataset(k, data=v.cpu().numpy())
+            y[k].attrs["__data_type__"] = "torch"
+        # ints and floats are stored as attributes of groups
+        elif isinstance(v, (int, float)):
+            y.attrs[k] = v
+        # resort to pickle for any other type of object
+        else:
+            int_data = np.fromstring(pickle.dumps(v), dtype="uint8")
+            y.create_dataset(k, data=int_data, dtype="uint8")
+            y[k].attrs["__data_type__"] = "pickle"
+
+
+def from_hdf5(
+    x: h5py.Group, device: Optional[str] = None
+) -> Hdf5ConvertibleType:
+    """Restore object from HDF5 group."""
+    # handle datasets
+    if isinstance(x, h5py.Dataset):
+        if x.attrs["__data_type__"] == "numpy":
+            y = np.empty(x.shape, dtype=x.dtype)
+            x.read_direct(y)
+        elif x.attrs["__data_type__"] == "torch":
+            y = torch.tensor(x, device=device)
+        elif x.attrs["__data_type__"] == "pickle":
+            y = pickle.loads(x[()])
+    # handle groups representing a dict or a batch
+    else:
+        y = {k: v for k, v in x.attrs.items() if k != "__convert_to__"}
+        for k, v in x.items():
+            y[k] = from_hdf5(v, device)
+        if "__convert_to__" in x. attrs:
+            # if dictionary represents Batch have to convert to Batch
+            if x.attrs["__convert_to__"] == "Batch":
```

Comment:
Why not just consider "Batch" as another `__data_type__`, instead of using a dedicated field `__convert_to__`.

Reply:
Ok, agreed. The code will be more consistent that way.